projects/core/src/product/services/product-page-meta.resolver.ts

```diff
@@ -179,4 +177,30 @@ export class ProductPageMetaResolver
   resolveRobots(): Observable<PageRobotsMeta[]> {
     return of([PageRobotsMeta.FOLLOW, PageRobotsMeta.INDEX]);
   }
+
+  /**
+   * Resolves the canonical url for the product page using the default canonical url
+   * configuration.
+   *
+   * In case of a variant product, the baseProduct code is used to resolve the url. It's important
+   * to know that this has a few limitations:
+   * - We're not always able to get the super baseProduct, in case of multi-level variants.
+   *   OCC only exposes the direct baseProduct, which might still not resolve in the correct
+   *   canonical URL. This is business driven and subject to change in a customization.
+   * - The url resolved for the variant doesn't contain any content other then the product code.
+   *   This means that we do not provide any product data to resolve pretty URLs (for example
+   *   the product title).
+   */
+  resolveCanonicalUrl(): Observable<string> {
+    return this.product$.pipe(
+      switchMap((product) => {
+        const code = product.baseProduct ?? product.code;
+        const url = this.routingService.getUrl({
+          cxRoute: 'product',
+          params: { code },
+        });
+        return this.pageLinkFactory?.resolveCanonicalUrl({}, url);
```

Comment:
[medium] We need a wrapper `of()`, because it's returned inside `switchMap()`. Otherwise it's working strange (it converts returned string to an array and emits its letters in separate emissions )
```suggestion
        return of(this.pageLinkFactory?.resolveCanonicalUrl({}, url));
```

Reply:
I noticed that same and moved to a map operation instead.